larray/core/session.py

```diff
@@ -1475,6 +1492,199 @@ def display(k, v, is_metadata=False):
         return res
 
 
+# the implementation of the class below is inspired by the 'ConstrainedBytes' class
+# from the types.py module of the 'pydantic' library
+class ConstrainedArrayImpl(Array):
+    expected_axes: AxisCollection
+
+    # see https://pydantic-docs.helpmanual.io/usage/types/#classes-with-__get_validators__
+    @classmethod
+    def __get_validators__(cls):
+        # one or more validators may be yielded which will be called in the
+        # order to validate the input, each validator will receive as an input
+        # the value returned from the previous validator
+        yield cls.validate
+
+    @classmethod
+    def validate(cls, value, field: ModelField):
+        # XXX: not sure we should accept a scalar as value but this allows to not repeat axes
+        # when setting a (default) value to the associated array
+        if not (isinstance(value, Array) or np.isscalar(value)):
+            raise TypeError("Expected object of type '{}' or a scalar for the variable '{}' "
+                            "but got object of type '{}'".format(Array.__name__, field.name, type(value).__name__))
+        # convert scalar to array with defined axes
+        if not isinstance(value, Array):
+            value = full(axes=cls.expected_axes, fill_value=value)
+        # check that passed array has compatible axes
+        else:
+            # XXX: we do not use 'cls.expected_axes == value.axes' because we consider
+            # that the order of axes is not important
+            try:
+                # XXX: should we allow broadcasting?
+                # if yes, the lines below must be replaced by something like:
+                #    cls.expected_axes.check_compatible(value.axes)
+                #    v = zeros_like(cls.expected_axes)
+                #    v[:] = value
+                #    value = v
+                if len(value.axes) != len(cls.expected_axes):
+                    raise Exception()
+                cls.expected_axes.check_compatible(value.axes)
+            except Exception:
+                msg = "Array '{}' was declared with axes \n{!r} " \
+                      "but got array with axes \n{!r}".format(field.name, cls.expected_axes, value.axes)
+                raise ValueError(msg)
+        return value
+
+
+# the implementation of the function below is inspired by the 'conbytes' function
+# from the types.py module of the 'pydantic' library
+def ConstantAxesArray(axes) -> Type[ConstrainedArrayImpl]:
+    if axes is not None and not isinstance(axes, AxisCollection):
+        axes = AxisCollection(axes)
+    namespace = {'expected_axes': axes}
+    return type('ConstrainedArrayImpl', (ConstrainedArrayImpl,), namespace)
```

Comment:
Can we avoid giving the subclass the same name than its parent?

And isn't this another way to spell:
```python
if axes is not None and not isinstance(axes, AxisCollection):
    axes = AxisCollection(axes)
class ConstrainedArrayImpl(ConstrainedArrayImpl):
    expected_axes = axes
return ConstrainedArrayImpl
```
**If** so, I find it more readable using the class keyword.

Reply:
I guess what you would like to test is:
```python
if axes is not None and not isinstance(axes, AxisCollection):
    axes = AxisCollection(axes)
class ConstrainedArray(ConstrainedArrayImpl):
    expected_axes = axes
return ConstrainedArray
```
?
Good question but since the class ``ConstrainedArray`` is defined inside the function in this case, I wonder how PyCharm will react with auto-completion? 
I'll try.

Reply:
There are two questions in there: 1) the name of the generated class (using the same name might indeed help PyCharm) and 2) can we use the class keyword which I find more readable than the code using type() (AFAICT, it is equivalent but I could be wrong).